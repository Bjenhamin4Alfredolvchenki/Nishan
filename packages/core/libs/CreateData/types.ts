import { ICache } from '@nishans/cache';
import { IOperation, TTextFormat } from '@nishans/types';
import { Logger } from '../../types';

export interface ParentCollectionData {
	parent_collection_id: string;
	name: TTextFormat;
	token: string;
	logger?: Logger;
	stack: IOperation[];
	cache: ICache;
	parent_relation_schema_unit_id: string;
}
