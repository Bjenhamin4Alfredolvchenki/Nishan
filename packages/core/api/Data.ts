import { TDataType, TData, TBlock, ISpace, IUserRoot, ICollection, ISpaceView } from '@nishans/types';
import { NishanArg, RepositionParams, UpdateCacheManuallyParam, FilterTypes, UpdateTypes } from '../types';
import { Operation, warn, positionChildren, iterateAndUpdateChildren, iterateAndGetChildren, iterateAndDeleteChildren } from "../utils";
import Operations from "./Operations";
interface IterateAndGetOptions<T>{
  child_type: TDataType,
  child_ids: string[] | keyof T,
  multiple?: boolean,
}

interface IterateAndUpdateOptions<T> extends IterateAndGetOptions<T>{
  manual?:boolean
}

interface IterateAndDeleteOptions<T> extends IterateAndUpdateOptions<T>{
  child_path?: keyof T,
}

/**
 * A class to update and control data specific stuffs
 * @noInheritDoc
 */

export default class Data<T extends TData> extends Operations {
  id: string;
  type: TDataType;
  #init_cache = false;

  constructor(arg: NishanArg & { type: TDataType }) {
    super(arg);
    this.type = arg.type;
    this.id = arg.id;
    this.#init_cache = false;
  }

  protected getLastEditedProps() {
    return { last_edited_time: Date.now(), last_edited_by_table: "notion_user", last_edited_by_id: this.user_id }
  }

  protected updateLastEditedProps(data?:TData){
    const target = data ?? this.getCachedData() as any;
    target.last_edited_time = Date.now();
    target.last_edited_by_table = "notion_user";
    target.last_edited_by_id = this.user_id;
  }

  /**
   * Get the cached data using the current data id
   */
  getCachedData() {
    const data = this.cache[this.type].get(this.id);
    if ((data as any).alive === false)
      warn(`${this.type}:${this.id} has been deleted`);
    return data as T;
  }

  async updateCachedData(){
    await this.updateCacheManually([[this.id, this.type]])
  }

  /**
   * Delete the cached data using the id
   */
  protected deleteCachedData() {
    this.cache[this.type].delete(this.id);
  }

  /**
   * Adds the passed block id in the child container array of parent
   * @param $block_id id of the block to add
   * @param arg
   * @returns created Operation and a function to update the cache and the class data
   */
  protected addToChildArray(parent: TData, position: RepositionParams) {
    this.stack.push(positionChildren({ logger: this.logger, child_id: this.id, position, parent, parent_id: this.id, parent_type: this.type }))
  }

  /**
   * Update the cache of the data using only the passed keys
   * @param arg
   * @param keys
   */
  updateCacheLocally(arg: Partial<T>, keys: ReadonlyArray<(keyof T)>, appendToStack?: boolean) {
    appendToStack = appendToStack ?? true
    const parent_data = this.getCachedData(), data = arg;

    Object.entries(arg).forEach(([key, value])=>{
      if(keys.includes(key as keyof T))
        parent_data[key as keyof T] = value;
    })

    this.logger && this.logger("UPDATE", this.type as any, this.id)
    if(appendToStack)
      this.stack.push(
        Operation[this.type].update(this.id,this.type === "user_settings" ? ["settings"] : [], data)
      );
  }

  protected async initializeCache() {
    if (!this.#init_cache) {
      const container: UpdateCacheManuallyParam = []
      if (this.type === "block") {
        const data = this.getCachedData() as TBlock;
        if (data.type === "page")
          container.push(...data.content);
        if (data.type === "collection_view" || data.type === "collection_view_page") {
          data.view_ids.map((view_id) => container.push([view_id, "collection_view"]))
          container.push([data.collection_id, "collection"])
        }
      } else if (this.type === "space") {
        const space = this.getCachedData() as ISpace;
        container.push(...space.pages);
        space.permissions.forEach((permission) => container.push([permission.user_id, "notion_user"]))
      } else if (this.type === "user_root")
        (this.getCachedData() as IUserRoot).space_views.map((space_view => container.push([space_view, "space_view"]))) ?? []
      else if (this.type === "collection") {
        container.push(...((this.getCachedData() as ICollection).template_pages ?? []))
        await this.queryCollection({
          collectionId: this.id,
          collectionViewId: "",
          query: {},
          loader: {
            type: "table",
            loadContentCover: true
          }
        }, 0)
      }
      else if (this.type === "space_view")
        container.push(...(this.getCachedData() as ISpaceView).bookmarked_pages ?? [])

      const non_cached = this.returnNonCachedData(container)
      await this.updateCacheManually(non_cached, 0);

      // If the block is a page, for all the collection block contents, fetch the collection attached with it as well
      if(this.type === "block"){
        const data = this.getCachedData() as TBlock;
        if(data.type === "page"){
          const collection_blocks_ids: UpdateCacheManuallyParam = [];
          for (let index = 0; index < data.content.length; index++) {
            const content_id = data.content[index];
            const content = this.cache.block.get(content_id)
            if(content && (content.type === "collection_view_page" || content.type === "collection_view"))
              collection_blocks_ids.push([content.collection_id, "collection"])
          }
          const non_cached = this.returnNonCachedData(collection_blocks_ids)
          await this.updateCacheManually(non_cached, 0);
        }
      }

      this.#init_cache = true;
    }
  }

  protected async deleteIterate<TD>(args: FilterTypes<TD>, options: IterateAndDeleteOptions<T>, transform: ((id: string) => TD | undefined), cb?: (id: string, data: TD) => void | Promise<any>) {
    await this.initializeCache()
    return  await iterateAndDeleteChildren<T, TD>(args, transform, {
      parent_id: this.id,
      parent_type: this.type,
      ...this.getProps(),
      ...options
    }, cb);
  }

  protected async updateIterate<TD, RD>(args: UpdateTypes<TD, RD>, options: IterateAndUpdateOptions<T>, transform: ((id: string) => TD | undefined), cb?: (id: string, data: TD, updated_data: RD) => any) {
    await this.initializeCache()
    return await iterateAndUpdateChildren<T, TD, RD>(args, transform, {
      parent_type: this.type,
      parent_id: this.id,
      ...this.getProps(),
      ...options
    }, cb);
  }

  protected async getIterate<RD>(args: FilterTypes<RD>, options: IterateAndGetOptions<T>, transform: ((id: string) => RD | undefined), cb?: (id: string, data: RD) => any) {
    await this.initializeCache()
    return await iterateAndGetChildren<T,RD>(args, transform, {
      parent_id: this.id,
      parent_type: this.type,
      ...this.getProps(),
      ...options,
    }, cb);
  }

  getProps() {
    return {
      token: this.token,
      interval: this.interval,
      user_id: this.user_id,
      shard_id: this.shard_id,
      space_id: this.space_id,
      cache: this.cache,
      logger: this.logger,
      stack: this.stack
    }
  }
}