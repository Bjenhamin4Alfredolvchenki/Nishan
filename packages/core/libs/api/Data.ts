import { NotionCache } from '@nishans/cache';
import { INotionFabricatorOptions } from '@nishans/fabricator';
import { NotionLogger } from '@nishans/logger';
import { NotionOperationPluginFunction, NotionOperations } from '@nishans/operations';
import { FilterTypes, IterateAndDeleteOptions, IterateAndGetOptions, IterateAndUpdateOptions, NotionTraverser, UpdateTypes } from "@nishans/traverser";
import { ICache, NotionCacheInitializerTracker, TData, TDataType } from '@nishans/types';
import { NotionUtils } from '@nishans/utils';
import { INotionCoreOptions } from '../';

/**
 * A class to update and control data specific stuffs
 * @noInheritDoc
 */

export default class NotionData<T extends TData> {
  id: string;
  type: TDataType;
  logger: boolean;
  user_id: string;
  notion_operation_plugins: NotionOperationPluginFunction[];
  shard_id: number;
  space_id: string
  interval: number;
  token: string;
  cache: ICache;
  cache_init_tracker: NotionCacheInitializerTracker;

  constructor(arg: INotionCoreOptions & { cache?: ICache, type: TDataType }) {
    this.type = arg.type;
    this.id = arg.id;
    this.logger = arg.logger ?? true;
    this.user_id = arg.user_id;
    this.notion_operation_plugins = arg.notion_operation_plugins ?? [];
    this.shard_id = arg.shard_id;
    this.space_id = arg.space_id;
		this.interval = arg.interval ?? 500;
		this.cache = (arg.cache && NotionCache.validateCache(arg.cache)) || NotionCache.createDefaultCache();
    this.cache_init_tracker = arg.cache_init_tracker || NotionCache.createDefaultCacheInitializeTracker();
    if(!arg.token)
      throw new Error(`Token not provided`);
		this.token = arg.token;
  }

  protected getLastEditedProps() {
    return { last_edited_time: Date.now(), last_edited_by_table: "notion_user", last_edited_by_id: this.user_id }
  }

  protected updateLastEditedProps(data?:TData){
    return NotionUtils.updateLastEditedProps(data ?? this.getCachedData() as any, this.user_id)
  }

  /**
   * Get the cached data using the current data id
   */
  getCachedData() {
    const data = this.cache[this.type].get(this.id);
    if (!data)
      NotionLogger.method.warn(`${this.type}:${this.id} doesnot exist in the cache`);
    else if((data as any).alive === false)
      NotionLogger.method.warn(`${this.type}:${this.id} is not alive`);
    return data as T;
  }

  async updateCachedData(){
    await NotionCache.updateCacheManually([[this.id, this.type]], this.getProps())
  }

  /**
   * Delete the cached data using the id
   */
  protected deleteCachedData() {
    this.cache[this.type].delete(this.id);
  }

  async updateCacheLocally(arg: Partial<T>, keys: ReadonlyArray<(keyof T)>) {
    const parent_data = this.getCachedData(), data = arg;

    Object.entries(arg).forEach(([key, value])=>{
      if(keys.includes(key as keyof T))
        parent_data[key as keyof T] = value;
      else
        delete (data as any)[key]
    })

    this.logger && NotionLogger.method.info(`UPDATE ${this.type} ${this.id}`);
    await NotionOperations.executeOperations([NotionOperations.Chunk[this.type].update(this.id, [], data)], this.getProps())
  }

  async initializeCacheForThisData() {
    if (this.cache_init_tracker[this.type].get(this.id) !== true && this.type !== "notion_user") {
      await NotionCache.initializeCacheForSpecificData(this.id, this.type, this.getProps())
    }
  }

  protected async deleteIterate<TD, C = any[]>(args: FilterTypes<TD>, options: IterateAndDeleteOptions<T, C>, transform: ((id: string) => TD | undefined | Promise<TD | undefined>), cb?: (id: string, data: TD) => void | Promise<any>) {
    if(options?.initialize_cache ?? true) await this.initializeCacheForThisData();
    return await NotionTraverser.delete<T, TD, C>(args, transform, {
      parent_id: this.id,
      parent_type: this.type,
      ...this.getProps(),
      ...options
    }, cb);
  }

  protected async updateIterate<TD, RD, C = any[]>(args: UpdateTypes<TD, RD>, options: IterateAndUpdateOptions<T, C>, transform: ((id: string) => TD | undefined | Promise<TD | undefined>), cb?: (id: string, data: TD, updated_data: RD, container: C) => any) {
    if(options?.initialize_cache ?? true) await this.initializeCacheForThisData();
    return await NotionTraverser.update<T, TD, RD, C>(args, transform, {
      parent_type: this.type,
      parent_id: this.id,
      ...this.getProps(),
      ...options
    }, cb);
  }

  protected async getIterate<RD, C>(args: FilterTypes<RD>, options: IterateAndGetOptions<T, C>, transform: ((id: string) => RD | undefined | Promise<RD | undefined>), cb?: (id: string, data: RD, container: C) => any) {
    if(options?.initialize_cache ?? true) await this.initializeCacheForThisData();
    return await NotionTraverser.get<T, RD, C>(args, transform, {
      parent_id: this.id,
      parent_type: this.type,
      ...this.getProps(),
      ...options,
    }, cb);
  }

  getProps() {
    return {
      token: this.token,
      interval: this.interval,
      user_id: this.user_id,
      shard_id: this.shard_id,
      space_id: this.space_id,
      cache: this.cache,
      logger: this.logger,
      notion_operation_plugins: this.notion_operation_plugins,
      cache_init_tracker: this.cache_init_tracker
    } as INotionFabricatorOptions
  }
}