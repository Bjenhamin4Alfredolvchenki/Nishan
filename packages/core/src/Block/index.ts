import CollectionViewPage from './CollectionViewPage';
import CollectionView from './CollectionView';
import Page from './Page';
import CollectionBlock from './CollectionBlock';
import Block from './Block';

export { Block, CollectionViewPage, CollectionView, Page, CollectionBlock };
