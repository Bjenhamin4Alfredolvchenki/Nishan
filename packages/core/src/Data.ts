import { TDataType, TData } from '@nishans/types';
import { NishanArg, RepositionParams, FilterTypes, UpdateTypes, Logger } from '../types';
import { Operation, warn, positionChildren, iterateAndUpdateChildren, iterateAndGetChildren, iterateAndDeleteChildren, constructLogger } from "../utils";
import Operations from "./Operations";

export interface IterateOptions<T, C>{
  /**
   * The data type of the child
   */
  child_type: TDataType,
  /**
   * A container of child ids or a key of the parent that stores the child ids
   */
  child_ids: string[] | keyof T,
  /**
   * Matches multiple based on the value
   */
  multiple?: boolean,
  /**
   * A container that stores the data
   */
  container: C
}

export interface IterateAndGetOptions<T, C> extends IterateOptions<T, C>{
}


export type IterateAndDeleteOptions<T, C> = IterateOptions<T, C> & ( {
  /**
   * Whether or not the user will manually handle all the mutations
   */
  manual: true
  /**
   * The key of the parent which contains the child ids
   */
  child_path?: keyof T,
} | {
  manual?: false
  child_path: keyof T,
})

export type IterateAndUpdateOptions<T, C> = IterateOptions<T, C> & {
  /**
   * Whether or not the user will manually handle all the mutations
   */
  manual?: boolean
}

/**
 * A class to update and control data specific stuffs
 * @noInheritDoc
 */

export default class Data<T extends TData> extends Operations {
  id: string;
  type: TDataType;
  init_cache = false;
  protected logger: Logger;
  user_id: string

  constructor(arg: NishanArg & { type: TDataType }) {
    super(arg);
    this.type = arg.type;
    this.id = arg.id;
    this.init_cache = false;
    this.logger = constructLogger(arg.logger);
    this.user_id = arg.user_id ?? '';
  }

  protected getLastEditedProps() {
    return { last_edited_time: Date.now(), last_edited_by_table: "notion_user", last_edited_by_id: this.user_id }
  }

  protected updateLastEditedProps(data?:TData){
    const target = data ?? this.getCachedData() as any;
    target.last_edited_time = Date.now();
    target.last_edited_by_table = "notion_user";
    target.last_edited_by_id = this.user_id;
  }

  /**
   * Get the cached data using the current data id
   */
  getCachedData() {
    const data = this.cache[this.type].get(this.id);
    if ((data as any).alive === false)
      warn(`${this.type}:${this.id} has been deleted`);
    return data as T;
  }

  async updateCachedData(){
    await this.updateCacheManually([[this.id, this.type]])
  }

  /**
   * Delete the cached data using the id
   */
  protected deleteCachedData() {
    this.cache[this.type].delete(this.id);
  }

  protected addToChildArray(parent_type: TDataType, parent: TData, position: RepositionParams) {
    this.stack.push(positionChildren({ logger: this.logger, child_id: this.id, position, parent, parent_type }))
  }

  /**
   * Update the cache of the data using only the passed keys
   * @param arg
   * @param keys
   */
  updateCacheLocally(arg: Partial<T>, keys: ReadonlyArray<(keyof T)>, appendToStack?: boolean) {
    appendToStack = appendToStack ?? true
    const parent_data = this.getCachedData(), data = arg;

    Object.entries(arg).forEach(([key, value])=>{
      if(keys.includes(key as keyof T))
        parent_data[key as keyof T] = value;
    })

    this.logger && this.logger("UPDATE", this.type as any, this.id)
    if(appendToStack)
      this.stack.push(
        Operation[this.type].update(this.id,this.type === "user_settings" ? ["settings"] : [], data)
      );
  }

  protected async initializeCacheForThisData() {
    if (!this.init_cache && this.type !== "notion_user") {
      await this.initializeCacheForSpecificData(this.id, this.type)
      this.init_cache = true;
    }
  }

  protected async deleteIterate<TD, C = any[]>(args: FilterTypes<TD>, options: IterateAndDeleteOptions<T, C>, transform: ((id: string) => TD | undefined), cb?: (id: string, data: TD) => void | Promise<any>) {
    await this.initializeCacheForThisData()
    return  await iterateAndDeleteChildren<T, TD, C>(args, transform, {
      parent_id: this.id,
      parent_type: this.type,
      ...this.getProps(),
      ...options
    }, cb);
  }

  protected async updateIterate<TD, RD, C = any[]>(args: UpdateTypes<TD, RD>, options: IterateAndUpdateOptions<T, C>, transform: ((id: string) => TD | undefined), cb?: (id: string, data: TD, updated_data: RD, container: C) => any) {
    await this.initializeCacheForThisData();
    return await iterateAndUpdateChildren<T, TD, RD, C>(args, transform, {
      parent_type: this.type,
      parent_id: this.id,
      ...this.getProps(),
      ...options
    }, cb);
  }

  protected async getIterate<RD, C>(args: FilterTypes<RD>, options: IterateAndGetOptions<T, C>, transform: ((id: string) => RD | undefined), cb?: (id: string, data: RD, container: C) => any) {
    await this.initializeCacheForThisData();
    return await iterateAndGetChildren<T, RD, C>(args, transform, {
      parent_id: this.id,
      parent_type: this.type,
      ...this.getProps(),
      ...options,
    }, cb);
  }

  getProps() {
    return {
      token: this.token,
      interval: this.interval,
      user_id: this.user_id ?? '',
      shard_id: this.shard_id,
      space_id: this.space_id,
      cache: this.cache,
      logger: this.logger,
      stack: this.stack
    }
  }
}