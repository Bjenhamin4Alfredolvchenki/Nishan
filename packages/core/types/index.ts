export * from './block';
export * from './maps';
export * from './nishan';
