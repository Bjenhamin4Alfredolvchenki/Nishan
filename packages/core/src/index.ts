import {NotionCacheClass} from "@nishans/cache";
import { INotionUser } from "@nishans/types";

import Block from "./Block";
import Collection from "./Collection";
import CollectionViewPage from "./CollectionViewPage";
import CollectionView from "./CollectionView";
import NotionUser from "./NotionUser";
import Page from "./Page";
import UserSettings from "./UserSettings";
import UserRoot from "./UserRoot";
import SpaceView from "./SpaceView";
import Space from "./Space";
import SchemaUnit from "./SchemaUnit";
import CollectionBlock from "./CollectionBlock";
import NotionPermissions from "./Permissions";
export * from "./View";
import { constructLogger, error, iterateAndGetChildren, transformToMultiple } from "../utils";
import {Logger, NishanArg,FilterType, FilterTypes} from "../types";

class Nishan extends NotionCacheClass {
  token: string;
  interval: number;
  init_cache: boolean;
  logger: Logger;

  constructor(arg: Pick<NishanArg, "token"> & { interval?: number, logger?: Logger }) {
    super(arg);
    this.token = arg.token;
    this.interval = arg.interval ?? 500;
    this.init_cache = false;
    this.logger = constructLogger(arg.logger)
  }

  #initializeCache = async () => {
    if (!this.init_cache) {
      await this.initializeNotionCache();
      this.init_cache = true;
    }
  }

  /**
   * Get `INotionUser` and return `NotionUser` by the `args` param
   * @param args An string id, a predicate passed the INotionUser or undefined to indicate everything
   */
  async getNotionUser(arg?: FilterType<INotionUser>) {
    return (await this.getNotionUsers(transformToMultiple(arg), false))[0];
  }

  /**
   * Get `INotionUser[]` and return `NotionUser[]` by the `args` param
   * @param args An array of string ids, a predicate passed the INotionUser or undefined to indicate everything
   */
  async getNotionUsers(args?: FilterTypes<INotionUser>, multiple?: boolean) {
    await this.#initializeCache();

    const common_props = {
      token: this.token,
      cache: this.cache,
      interval: this.interval,
      logger: this.logger,
      stack: [],
    }

    const notion_user_ids: string[] = [];
    for (const [notion_user_id] of this.cache.notion_user) {
      notion_user_ids.push(notion_user_id)
    }

    return await iterateAndGetChildren<INotionUser, INotionUser, NotionUser[]>(args, (id)=>this.cache.notion_user.get(id), {
      ...common_props,
      multiple,
      child_ids: notion_user_ids,
      child_type: 'notion_user',
      container: [],
      parent_id: '',
      parent_type: 'notion_user'
    }, (id, _, container)=>container.push(new NotionUser({ ...common_props, user_id: id, id, space_id: "0", shard_id: 0 })));
  }
}

export default Nishan;
export * from "../types";
export * from "../utils";

export {
  Block,
  Collection,
  CollectionViewPage,
  CollectionView,
  NotionUser,
  Page,
  UserSettings,
  UserRoot,
  SpaceView,
  Space,
  SchemaUnit,
  CollectionBlock,
  NotionPermissions
}