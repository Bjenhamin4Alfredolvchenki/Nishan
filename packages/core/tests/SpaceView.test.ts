import { ICache } from '@nishans/cache';
import { IOperation } from '@nishans/types';
import deepEqual from 'deep-equal';
import { SpaceView } from '../src';
import Data from '../src/Data';

afterEach(() => {
	jest.clearAllMocks();
});

it(`getCachedParentData`, () => {
	const cache: ICache = {
			block: new Map(),
			collection: new Map(),
			collection_view: new Map(),
			notion_user: new Map(),
			space: new Map(),
			space_view: new Map([ [ 'space_view_1', { alive: true } as any ] ]),
			user_root: new Map([ [ 'user_root_1', { space_views: [ 'space_view_1' ] } as any ] ]),
			user_settings: new Map()
		},
		stack: IOperation[] = [];

	const space_view = new SpaceView({
		cache,
		id: 'space_view_1',
		stack,
		interval: 0,
		shard_id: 123,
		space_id: 'space_1',
		token: 'token',
		user_id: 'user_root_1'
	});

	const parent_data = space_view.getCachedParentData();
	expect(deepEqual(parent_data, { space_views: [ 'space_view_1' ] })).toBe(true);
});

it(`reposition`, () => {
	const cache: ICache = {
			block: new Map(),
			collection: new Map(),
			collection_view: new Map(),
			notion_user: new Map(),
			space: new Map(),
			space_view: new Map([
				[
					'space_view_1',
					{
						bookmarked_pages: [ 'block_1' ]
					} as any
				]
			]),
			user_root: new Map([ [ 'user_root_1', { id: 'user_root_1', space_views: [ 'space_view_1' ] } as any ] ]),
			user_settings: new Map()
		},
		stack: IOperation[] = [];

	const space_view = new SpaceView({
		cache,
		id: 'space_view_2',
		stack,
		interval: 0,
		shard_id: 123,
		space_id: 'space_1',
		token: 'token',
		user_id: 'user_root_1'
	});
	const addToChildArrayMock = jest.spyOn(Data.prototype, 'addToChildArray' as any);

	space_view.reposition(0);
	expect(addToChildArrayMock).toHaveBeenCalledTimes(1);
	expect(addToChildArrayMock).toHaveBeenCalledWith(
		'user_root',
		{
			id: 'user_root_1',
			space_views: [ 'space_view_2', 'space_view_1' ]
		},
		0
	);
});

it(`update`, () => {
	const cache: ICache = {
			block: new Map(),
			collection: new Map(),
			collection_view: new Map(),
			notion_user: new Map(),
			space: new Map(),
			space_view: new Map(),
			user_root: new Map(),
			user_settings: new Map()
		},
		stack: IOperation[] = [];

	const space_view = new SpaceView({
		cache,
		id: 'space_view_1',
		stack,
		interval: 0,
		shard_id: 123,
		space_id: 'space_1',
		token: 'token',
		user_id: 'user_root_1'
	});

	const updateCacheLocallyMock = jest.spyOn(Data.prototype, 'updateCacheLocally').mockImplementationOnce(() => {
		return {} as any;
	});

	space_view.update({
		joined: false
	});

	expect(updateCacheLocallyMock).toHaveBeenCalledTimes(1);
	expect(updateCacheLocallyMock).toHaveBeenCalledWith(
		{
			joined: false
		},
		[ 'notify_desktop', 'notify_email', 'notify_mobile', 'joined', 'created_getting_started' ]
	);
});

it(`getSpace`, () => {
	const cache: ICache = {
			block: new Map(),
			collection: new Map(),
			collection_view: new Map(),
			notion_user: new Map(),
			space_view: new Map([ [ 'space_view_1', { id: 'space_view_1', space_id: 'space_1' } as any ] ]),
			space: new Map([ [ 'space_2', { id: 'space_2' } as any ], [ 'space_1', { id: 'space_1' } as any ] ]),
			user_root: new Map(),
			user_settings: new Map()
		},
		stack: IOperation[] = [];

	const logger_spy = jest.fn();

	const space_view = new SpaceView({
		cache,
		logger: logger_spy,
		id: 'space_view_1',
		stack,
		interval: 0,
		shard_id: 123,
		space_id: 'space_1',
		token: 'token',
		user_id: 'user_root_1'
	});

	const space = space_view.getSpace();
	expect(logger_spy).toHaveBeenCalledTimes(1);
	expect(logger_spy).toHaveBeenCalledWith('READ', 'space', 'space_1');
	expect(deepEqual(space.getCachedData(), { id: 'space_1' })).toBe(true);
});

it(`getBookmarkedPage`, async () => {
	const cache: ICache = {
			block: new Map([
				[ 'block_1', { type: 'collection_view_page', collection_id: 'collection_1', view_ids: [] } as any ]
			]),
			collection: new Map([ [ 'collection_1', { name: [ [ 'Collection One' ] ] } as any ] ]),
			collection_view: new Map(),
			notion_user: new Map(),
			space_view: new Map([ [ 'space_view_1', { id: 'space_view_1', bookmarked_pages: [ 'block_1' ] } as any ] ]),
			space: new Map(),
			user_root: new Map(),
			user_settings: new Map()
		},
		stack: IOperation[] = [];

	const logger_spy = jest.fn();

	const space_view = new SpaceView({
		cache,
		logger: logger_spy,
		id: 'space_view_1',
		stack,
		interval: 0,
		shard_id: 123,
		space_id: 'space_1',
		token: 'token',
		user_id: 'user_root_1'
	});

	const page_map = await space_view.getBookmarkedPage('block_1');
	expect(page_map.collection_view_page.get('Collection One')).not.toBeUndefined();
});

it(`updateBookmarkedPages`, async () => {
	const space_view_1 = { id: 'space_view_1', bookmarked_pages: [ 'block_1' ] },
		cache: ICache = {
			block: new Map([
				[ 'block_1', { type: 'page', properties: { title: [ [ 'Block One' ] ] } } as any ],
				[ 'block_2', { type: 'page', properties: { title: [ [ 'Block Two' ] ] } } as any ]
			]),
			collection: new Map(),
			collection_view: new Map(),
			notion_user: new Map(),
			space_view: new Map([ [ 'space_view_1', space_view_1 as any ] ]),
			space: new Map(),
			user_root: new Map(),
			user_settings: new Map()
		},
		stack: IOperation[] = [];

	const logger_spy = jest.fn();

	const space_view = new SpaceView({
		cache,
		logger: logger_spy,
		id: 'space_view_1',
		stack,
		interval: 0,
		shard_id: 123,
		space_id: 'space_1',
		token: 'token',
		user_id: 'user_root_1'
	});

	const page_map = await space_view.updateBookmarkedPages([ [ 'block_1', false ], [ 'block_2', true ] ]);

	expect(page_map.page.get('Block One')).not.toBeUndefined();
	expect(page_map.page.get('block_2')).not.toBeUndefined();

	expect(space_view_1).toMatchSnapshot({
		bookmarked_pages: [ 'block_2' ],
		id: 'space_view_1',
		last_edited_time: expect.any(Number),
		last_edited_by_table: 'notion_user',
		last_edited_by_id: 'user_root_1'
	});

	expect(stack.length).toBe(3);
});
