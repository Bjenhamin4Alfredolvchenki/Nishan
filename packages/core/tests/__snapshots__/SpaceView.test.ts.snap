// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`updateBookmarkedPages 1`] = `
Object {
  "bookmarked_pages": Array [
    "block_2",
  ],
  "id": "space_view_1",
  "last_edited_by_id": "user_root_1",
  "last_edited_by_table": "notion_user",
  "last_edited_time": Any<Number>,
}
`;
